### DISCUSSION TITLE
Generate type map for each model

### DISCUSSION THREAD

**Discussion by dever23b** [prisma/prisma#18322] [Q&A] [answered]
## Problem

Attempting to migrate to Prisma and having a tough time wrapping any features generically due to the (apparent?) lack of umbrella exposing the applicable types for a given model.  Not sure if I'm just doing this wrong, or if this is a legitimate request you may hopefully consider.

My ultimate goal is to be able to use generics to abstract some of the redundant code away from lifecycle operations in my project.  For some of my models, I want to have an audit log of changes.  When records are created/updated/deleted, I log these events and record the changes in a separate table.  

In my first attempt to accomplish this with Prisma, I read into the Client Extensions preview feature and wrote several extensions.  Unfortunately, this route didn't seem to be a good fit for my overall objective.  After reading through #15074 and the comments thereof, I found myself in agreement with the opposition comments to the feature.  I think what you do with Prisma is simple, focused, and effective and perhaps the CEs would be better implemented in other was in our own projects.  Accordingly, I set out to write a new layer of code to sit between the rest of my project and Prisma.  In addition to my auditing objective, I also have a myriad of computed fields and such on various models, so I opted to write a class for each model, to serve my goals.

Unfortunately, I'm having a really tough time being able to abstract anything from my classes due to the typings available in Prisma.  If I write everything in one class for a given model, it works great; however, this means I have to copy/paste quite a bit of redundant code between several other models, which is a recipe for failure as the code evolves.  What I would really like to be able to do is have an abstract, generic base class that can take a few type parameters and retain a lot of the reusable code and then extend this class with each model.  I can't find any way to do anything effective with the abstract class, though, because there doesn't seem to be any generic way to refer to the various model types generated by `prisma generate`: they're all independent of each other, with no base extension or type map.  The `Prisma.TypeMap` has been hugely helpful in many regards, but it doesn't help with exposing models' associated types.  For example, I can't find a way to dynamically access a given model's `GetPayload` or `Input` types.  I've been able to type the appropriate Prisma model from a generic using the code below, but the signature required is very redundant and I still haven't been able to dynamically access the other necessary types in order to type the generic methods I want to write.  

```typescript
import { Prisma, PrismaClient } from "@prisma/client";
export abstract class Model<Client extends Prisma.TypeMap['meta']['modelProps'], Payload extends { id: number }> {
  protected abstract _modelName: Prisma.ModelName;
  protected abstract _client: PrismaClient[Client];
  protected _values: Payload;
  
  protected constructor(data: Payload) {
    this._values = data;
  }


  public static async create(data: HowToTypeThis) {
    // this seems to be a dead end
  }
}
```

I've been brainstorming for a few days, so perhaps I am going about this all wrong and would certainly appreciate an alternative suggestion if I'm just lost in the woods and not finding trees.

## Suggested solution

What would be helpful is a map similar to `Prisma.TypeMap` that, rather than aggregating types by operations, aggregated the various Input, Args, Output, etc types under a key for each model.  Also, an association between the `camelCase` and `PascalCase` namings for each type, so that we can dynamically retrieve the appropriate Prisma Client and its types.

## Additional context

Here's some rough pseudocode I put together as an example of what I think should be a lot more easily possible.  
```typescript
// lib/Model.class.ts
import { Prisma, PrismaClient } from '@prisma/client'
export abstract interface Model<Name extends Prisma.ModelName, Payload extends Prisma.Models[Name]['payloads']['allIncludes']> {
  onCreate?: () => void | Promise<void>;
  onRead?: () => void | Promise<void>;
  onUpdate?: () => void | Promise<void>;
  onDelete?: () => void | Promise<void>;
}

export abstract class Model<Name extends Prisma.ModelName, Payload extends Prisma.Models[Name]['payloads']['allIncludes']> {
  protected abstract _client: PrismaClient[Prisma.Models[Name]['client']];
  protected _data: {
    stored: Payload,
    current: Payload,
  }
  protected _dirty: string[] = [];

  protected constructor(data: Payload) {
    this._data.current = data;
    this._data.stored = data;

    for( let key in data ) {
      Object.defineProperty(this, key, {
        get: () => this._data.current[key],
        set: (value: unknown) => {
          if( this._data.stored[key] !== value ) {
            this._dirty.push(key);
          }
          if( this._data.current[key] !== value ) {
            this._data.current[key] = value;
          }
        }
      })
    }
  }

  protected _recalculateDirty() {
    const dirty: string[] = [];
    for( const key in Object.keys(this._data.stored) ) {
      const oldValue = this._data.stored[key as keyof Payload];
      const newValue = this._data.current[key as keyof Payload];
      if( oldValue !== newValue) {
        dirty.push(key);
      }
    }
    this._dirty = dirty;
  }

  public reload(data?: Payload) {
    if( data === undefined ) {
      data = this._client.findFirst({ where: { id: this._data.current.id }});
    }
    this._data.stored = data;
    this._recalculateDirty();
    return data;
  }

  public reset() {
    this._data.current = this._data.stored;
    this._dirty = [];
  }

  public save() {
    if( this.onUpdate ) this.onUpdate();

    return this._client
      .update({
        data: this._data.current,
        where: { id: this._data.current.id },
      })
      .then((record) => this.reload(record))
  }

  public delete() {
    if( this.onDelete ) this.onDelete();
    return this._client.delete({ where: { id: this._data.current.id }})
  }
}
```
```typescript
// lib/ModelWithAuditing.class.ts
import { Prisma, PrismaClient } from '@prisma/client'
import { Model } from './Model.class';
export abstract class ModelWithAudit<Name extends Prisma.ModelName, Payload extends Prisma.Models[Name]['payloads']['allIncludes']> extends Model<Name, Payload> {
  async onCreate() {
    // make entry into audit log
  }
  async onUpdate() {
    // make entry into audit log
  }
  async onDelete() {
    // make entry into audit log
  }
}
```
```typescript
// models/User.class.ts
import { ModelWithAudit } from '../lib/ModelWithAudit.class';
class User extends ModelWithAudit<'user'> {
  onDelete() {
    if( logicThatFindsAnyFlags ) {
      throw new UserCannotBeDeletedException();
    }
    super.onDelete()
  }

  get fullName() { return `${this.firstName} ${this.lastName}`}
}
```

**Comment by nurul3101** [comment, marked as answer]
Hello @dever23b 👋 

First of all, thank you so much for taking the time to construct such a detailed write-up.
We have a few feature requests which I believe would help you in this use case.

- https://github.com/prisma/prisma/issues/12604
- https://github.com/prisma/prisma/issues/2506

Can you have a look at these feature requests?

**Reply by nurul3101** [reply]
@dever23b Do let us know in case you have any further queries or feedback

**Reply by dever23b** [reply]
I'm so sorry, somehow the notification for this topic fell through the cracks and I didn't realize anyone had replied here.  

As I understand the noted requests, neither of them (if implemented) would facilitate what I would like to accomplish.  

Both of these requests seem to be focused on modifying the Prisma Schema file syntax/language to support things like "extends" and "implements" similar to JavaScript/TypeScript/etc.  I like both ideas; however, they are focused on a different aspect of Prisma than my request.

While those requests are focused on the Prisma Schema file, my request is focused on TypeScript functionality and being able to wrap the models generated from the Schema file into classes that can contain additional functionality for the models.  What I want to be able to do is implement a base, abstract, class in TypeScript and then extend it with classes that represent the Prisma models, retain Prisma's functionality and retain Prisma's type safety.

The example I wrote above illustrates two features I am trying to add to models in my project: hooks and change tracking.  While I didn't include all of the necessary code in the example, the idea is that a base class facilitates storing both the original retrieved values from the database, as well as a "current" set of values.  The base class exposes a method that can return an aggregation of any values that do not match what is contained in the database.  It also facilitates a hook mechanism, so that my models can have, as an example, an "onSave" hook that inspects any changes and logs those specific changes to an audit log.

I don't think any of this functionality I am writing needs to be built into Prisma: I think it's outside of the scope of what Prisma is focused on and excels at.  However, it would be very helpful if Prisma (specifically, the `prisma generate` codegen that creates all of the types and suchforth based on the Schema) exposed just a little bit more to allow us to create such a class system.

I hope this makes sense.  I'm happy to answer questions, and I certainly apologize for the delay.

**Reply by dever23b** [reply]
If it helps at all, the "why" behind this is that my project will routinely, automatically, import a data file, and then parse that data into several different models for storage.  The functionality I am trying to write is designed to keep an eye on the automated imports and maintain a change log (implemented as a separate DB model) for troubleshooting purposes.  

In the interest of organizing code and keeping relevant tasks together, it seems like the most appropriate place to implement this tracking and logging would be on or as close to the model itself.  Since Prisma does not provide this functionality (nor do I think it needs to on its own), my idea was to create a class system to sit on top of the Prisma models & data.  The classes offer a myriad of other potential functionality too, as suggested in your documentation ("Custom Models").  Indeed, some of my model classes also have getters that resolve properties not actually stored in the DB.  This class system works great, individually.  However, without the ability to abstract some of the common code and then extend it, it's creating a nightmare of maintaining redundant code copied between dozens of classes.  All of my attempts so far to extract any common code into a base class have failed to maintain type safety and ended up with a bunch of TypeScript compile errors.  

It seems like if the various model types were declared as interfaces and extended a common interface, I could then refer to that common interface with my base class and have things work out properly.

As an example, when I've tried to store a reference to the Prisma client, and I type it with `PrismaClient[Client]` (in the original example), it's very difficult to use any of the methods on that client reference without running into TypeScript errors because the type resolves to a massive union of every potential client type and they don't seem (to TS) to have anything in common with each other.

**Reply by dever23b** [reply]
My goal is to be able to write a generic class, in TypeScript, that can successfully resolve to a Prisma model (based on what gets generated via `prisma generate`) and be able to resolve all of the necessary types for that model  (ie: the 'Where', 'Select', 'Input', etc utility types) in order to retain Prisma's functionality.

I want to do something very close to your examples in the Custom Models page of the documentation; however, I'm trying to write this as a generic class that can be extended instead of having to copy/paste the same code between a dozen models and try to keep that maintained successfully.

**Reply by nurul3101** [reply]
Thank you so much for taking the time to explain the functionality you would like to achieve.

> It would be very helpful if Prisma (specifically, the prisma generate codegen that creates all of the types and suchforth based on the Schema) exposed just a little bit more to allow us to create such a class system.

I would recommend creating a [feature request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) with the missing types/methods that the `generate` command should create.
